### PR TITLE
Eip1283

### PR DESCRIPTION
**Pull request by holiman**
This PR implements [EIP 1283(https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1283.md#specification-version-ii). 1283 is an alternative to 1087, which does not require as much post-processing.
 
The cost for SSTORE operations rely on current value, new value and `original` value, where `original` means 'what is currently in the trie' a.k.a 'the value that would remain if this current transaction was wholly reverted'. 

This PR uses a map in state_object to store original value whenever a slot is updated. 
This PR is experimental, in that it does not contain any testcases whatsoever, the intention being more to test how complex it would be to implement rather than being a fully-fledged merge-ready PR.

### CODE REVIEW

**Comment by holiman** [comment]
Thanks for checking, @AlexeyAkhunov ! I had a feeling that the implementation could be done better, by making better use of the existing structures, but didn't really go in for it. If we decide to move forward with this EIP, I'll update it according to your observations.

**Comment by holiman** [comment]
(or, actually, might be better just rewrite it based on @karalabe's PR instead)

File: core/vm/gas_table.go [hunk 1]
```diff
@@ -116,79 +116,70 @@ func gasReturnDataCopy(gt params.GasTable, evm *EVM, contract *Contract, stack *
...
-		// 1. current == new
-		return 200, nil
-	}
-	original := evm.StateDB.GetStateOriginal(contract.Address(), common.BigToHash(x))
-	// 2
-	if original == current { // 2.1
-		if original == (common.Hash{}) { // 2.1.1
-			return 20000, nil
+	// The legacy gas metering only takes into consideration the current state
+	if !evm.chainRules.IsConstantinople {
```
Comment (holiman) [inline core/vm/gas_table.go]: I personally think it was cleaner to have two separate gas calc functions, one enabled in Constantinople and one earlier. That makes the respective methods smaller, and saves this check on every `SSTORE`.

**Comment by holiman** [comment]
A lot of the tests are failing, since this PR messes with a lot of things that use `SSTORE`.  I'll take a look and see if those can be updated.

**Comment by holiman** [comment]
See https://github.com/ethereum/tests/pull/511#issuecomment-422589402 -- before all tests are regenerated, not much we can do to turn travis green ...

File: core/vm/gas_table.go (hunk 1 above)
Comment (karalabe) [reply core/vm/gas_table.go]: It's a single bool comparison vs. an sstore potentially loads of data from disk. Other gas calculations (e.g. `gasSuicide`) also use the same logic of having a single method per opcode and potentially checking whether a fork is enabled or not. It feels a bit cleaner for me to have a single method, but if you really disagree we can split it.

Comment (holiman) [reply core/vm/gas_table.go]: Yeah, it's not a heavy toll, the main reason for me was more that it allowed splitting the methods into two separate units. 
I don't feel strongly about it, you're the main maintainer :)

**Comment by karalabe** [comment]
> See ethereum/tests#511 (comment) -- before all tests are regenerated, not much we can do to turn travis green ...

Perhaps we should disable the Constantinople tests and reenable them when they're updated. I'd very much like to keep pre-Constantinople tests live in our repo.

**Comment by holiman** [comment]
I can fix that

**Pull request merged**